### PR TITLE
Improvements for Gurobi/Cplex Interfaces (mostly fixing tests and switching to bulk extraction of solution values)

### LINKED ISSUE #327: Direct and persistent interfaces to Gurobi and Cplex do not load slacks for range contraints
This is breaking tests.

### PR DESCRIPTION
**Pull request by michaelbynum**
## Summary/Motivation:
1) The slacks for range constraints was causing many tests to fail.
2) Bulk extraction of variable values, duals, etc. from Cplex/Gurobi is more efficient - especially for those using Gurobi cloud licenses.

## Changes proposed in this PR:
1) The direct/persistent solver interfaces now report slacks for range constraints. 
2) I converted much of the code that loads solutions from Gurobi/Cplex to extract variable values, duals, etc. in bulk rather than one at a time. This was done for the Cplex direct/persistent interfaces, the Gurobi direct/persistent interfaces, and for the Gurobi lp file interface.

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by jsiirola** [comment]
@ghackebeil: is there any chance that you could review this PR?  You are probably the one most familiar with this part of the code...

File: pyomo/solvers/plugins/solvers/GUROBI_RUN.py [hunk 1]
```diff
@@ -258,25 +258,39 @@ def gurobi_run(model_file, warmstart_file, soln_file, mipgap, options, suffixes)
         solnfile.write('objective: %s\n' % str(obj_value))
         solnfile.write('gap: 0.0\n')
 
-        for var in vars:
-            solnfile.write('var: %s : %s\n' % (str(var.getAttr(GRB.Attr.VarName)), str(var.getAttr(GRB.Attr.X))))
+        vals = model.getAttr("X", vars)
+        names = model.getAttr("VarName", vars)
+        for val, name in zip(vals, names):
```
Comment (ghackebeil) [inline pyomo/solvers/plugins/solvers/GUROBI_RUN.py]: I recommend adding `from six.moves import zip` for better Py2 performance

File: pyomo/solvers/plugins/solvers/direct_or_persistent_solver.py [hunk 2]
```diff
@@ -276,3 +278,58 @@ def _get_version(self):
...
+        return self._data[id(key)]
+
+    def __setitem__(self, key, value):
+        self._data[id(key)] = value
+        self._reversed[id(value)] = key
+
+    def __delitem__(self, key):
+        val = self._data[id(key)]
+        del self._data[id(key)]
+        del self._reversed[id(val)]
```
Comment (ghackebeil) [inline pyomo/solvers/plugins/solvers/direct_or_persistent_solver.py]: `val` is undefined in this method.

As a broader point, I am opposed to implementing a class like this for such a narrow use case (unless it is fully unit tested). Two separate ComponentMap() objects would be sufficient. I don't want to say this should hold up the release. It is just something to consider.

File: pyomo/solvers/plugins/solvers/GUROBI_RUN.py (hunk 1 above)
Comment (michaelbynum) [reply pyomo/solvers/plugins/solvers/GUROBI_RUN.py]: @ghackebeil Good suggestion. Thanks. I will update this.

File: pyomo/solvers/plugins/solvers/direct_or_persistent_solver.py (hunk 2 above)
Comment (michaelbynum) [reply pyomo/solvers/plugins/solvers/direct_or_persistent_solver.py]: @ghackebeil That is fair. I was being lazy. I'll get this updated as well.

**Comment by michaelbynum** [comment]
Gabe approved the pull request, so I am going to merge it and close the associated issues.

**Pull request merged**